PARCEL_2_RFC.md

```diff
@@ -0,0 +1,1285 @@
+
+When you do need to configure Parcel, it will be in one of 3 places.
+
+- If you need to configure the CLI, it will be a CLI flag
+- If you need to configure your package, it will be in the `package.json`
+- If you need to configure something with your files or the Parcel asset
+  pipeline, it will be in `.parcelrc`
+
+### `package.json`
+
+[todo]
+
+```json
+{
+  "name": "foo",
+  "main": "dist/main/index.js",
+  "module": "dist/module/index.js",
+  "browser": "dist/browser/index.js",
+  "browserslist": ["> 1%", "not dead"],
+  "engines": {
+    "node": ">=4.x"
+  },
+  "source": "src/index.js",
+  "targets": {
+    "main": {
+      "node": ["^4.0.0"]
+    },
+    "module": {
+      "node": ["^8.0.0"]
+    },
+    "browser": {
+      "browsers": ["> 1%", "not dead"]
+    }
+  },
+  "alias": {
+    "react": "preact-compat",
+    "react-dom": "preact-compat"
+  }
+}
+```
+
+#### `package.json#name`
+
+**(Required)** The name of the package is always required in order to be
+considered a valid `package.json`.
+
+```json
+{
+  "name": "my-package"
+}
+```
+
+#### `package.json#version`
+
+**(Required)** All packages inside `node_modules` must have a `package.json#version`.
+
+```json
+{
+  "version": "1.0.0"
+}
+```
+
+#### `package.json#main`
+
+**(Required)** This is the "main" target's entry point for the package.
+
+```json
+{
+  "main": "dist/main/index.js"
+}
+```
+
+See [Targets](#targets)
+
+#### `package.json#module`
+
+This is the "module" target's entry point for the package.
+
+```json
+{
+  "module": "dist/module/index.js"
+}
+```
+
+See [Targets](#targets)
+
+#### `package.json#browser`
+
+This is the "browser" target's entry point for the package.
+
+```json
+{
+  "browser": "distflinesof/browser/index.js"
+}
+```
+
+See [Targets](#targets)
+
+#### `package.json#source`
+
+Specify the entry points for your source code which gets mapped to your
+targets.
+
+```json
+{
+  "source": "src/index.js",
+  "source": ["src/index.js", "src/index.html"]
+}
+```
+
+See [Sources](#sources)
+
+#### `package.json#browserslist`
+
+As specified by Browserslist, this field is for specifying which transforms
+should be applied to browser bundles.
+
+```json
+{
+  "browserslist": ["> 0.2%", "not dead"]
+}
+```
+
+See [Environments](#environments)
+
+#### `package.json#engines`
+
+Specify what versions of what engines you want to support.
+
+```json
+{
+  "engines": {
+    "node": ">=4.x",
+    "electron": ">=2.x"
+  }
+}
+```
+
+See [Environments](#environments)
+
+#### `package.json#targets`
+
+Configuration for individual targets.
+
+```json
+{
+  "targets": {
+    "main": {
+      "engines": {
+        "node": ">=4.x",
+        "electron": ">=2.x"
+      },
+      "browsers": ["> 1%", "not dead"]
+    }
+  }
+}
+```
+
+See [Targets](#targets)
+
+#### `package.json#alias`
+
+Aliases asset names/paths to other assets.
+
+```json
+{
+  "alias": {
+    "react": "preact-compat",
+    "react-dom": "preact-compat"
+  }
+}
+```
+
+See [Aliases](#aliases)
+
```

Comment:
Should we support externals? #144

Reply:
It's a good question, I think we need to talk about it a bit more.

One part of the problem I think belongs in aliases:

```json
"aliases": {
  "react": "https://unpkg.com/react@16.4.2/umd/react.production.min.js"
}
```

Reply:
Awesome, however I'm not sure how that would work in practice, as the interface on that is different than the node_modules version.

My main concern is probably that there is not really a standard for how to publish prebuilt packages.

Reply:
@mohsen1 That looks great to me. Ideally Parcel could handle falling back to a local module if the external one fails to load, rather than having to manually write the code for that (see my comment in https://github.com/parcel-bundler/parcel/issues/144) but perhaps that's outside the scope of Parcel?

Reply:
Will this work with out a url? In case you want to replace a module (@wordpress/blob) with a value (wp.blob) to be used to retrieve a global (wp.blob) variable. 

example from webpack
```js 
{ 
  externals: { 
    '@wordpress/blob': 'wp.blob'
   }
}
```

Or will this type of behavior not be supported by aliases?